libpod/lock/lock.go

```diff
@@ -0,0 +1,55 @@
+package lock
+
+// Manager provides an interface for allocating multiprocess locks.
+// Locks returned by Manager MUST be multiprocess - allocating a lock in
+// process A and retrieving that lock's ID in process B must return handles for
+// the same lock, and locking the lock in A should exclude B from the lock until
+// it is unlocked in A.
+// All locks must be identified by a UUID (retrieved with Locker's ID() method).
+// All locks with a given UUID must refer to the same underlying lock, and it
+// must be possible to retrieve the lock given its UUID.
+// Each UUID should refer to a unique underlying lock.
+// Calls to AllocateLock() must return a unique, unallocated UUID.
+// AllocateLock() must fail once all available locks have been allocated.
+// Locks are returned to use by calls to Free(), and can subsequently be
+// reallocated.
+type Manager interface {
+	// AllocateLock returns an unallocated lock.
+	// It is guaranteed that the same lock will not be returned again by
+	// AllocateLock until the returned lock has Free() called on it.
+	// If all available locks are allocated, AllocateLock will return an
+	// error.
+	AllocateLock() (Locker, error)
+	// RetrieveLock retrieves a lock given its UUID.
+	// The underlying lock MUST be the same as another other lock with the
+	// same UUID.
+	RetrieveLock(id string) (Locker, error)
+}
+
+// Locker is similar to sync.Locker, but provides a method for freeing the lock
+// to allow its reuse.
+// All Locker implementations must maintain mutex semantics - the lock only
+// allows one caller in the critical section at a time.
+// All locks with the same ID must refer to the same underlying lock, even
+// if they are within multiple processes.
+type Locker interface {
+	// ID retrieves the lock's ID.
+	// ID is guaranteed to uniquely identify the lock within the
+	// Manager - that is, calling RetrieveLock with this ID will return
+	// another instance of the same lock.
+	ID() string
+	// Lock locks the lock.
+	// This call MUST block until it successfully acquires the lock or
+	// encounters a fatal error.
+	Lock() error
+	// Unlock unlocks the lock.
+	// A call to Unlock() on a lock that is already unlocked lock MUST
+	// error.
+	Unlock() error
+	// Deallocate deallocates the underlying lock, allowing its reuse by
+	// other pods and containers.
+	// The lock MUST still be usable after a Free() - some libpod instances
+	// may still retain Container structs with the old lock. This simply
+	// advises the manager that the lock may be reallocated.
```

Comment:
The problem this is trying to solve: processes who have a copy of the container struct won't realize it was deleted until they perform a (locked) `syncContainer()` to check the database for updates. We need to be able to grab the lock for this brief check that the container still exists even after it has been deleted.
This should be fine as the only time we take two container locks simultaneously internally is when one container depends on another, and we have a number of checks to ensure that a container cannot depend on a removed container and a dependency container cannot be removed before the containers that depend on it - so we shouldn't ever reuse a lock in that case.

I don't really like this, but I can't think of another way of doing this without expensive reference counting for active instances of a lock in a container struct, which could easily become stale and prevent reallocation.

Reply:
> We need to be able to grab the lock for this brief check that the container still exists even after it has been deleted.

(I still don’t overall quite understand what the locks are going to be used for (so most of my comments may well be completely off-base). Just relying on the database transactions to read the container status should give exactly the same guarantee — the containers has existed at the time the {DB,lock} was locked, and nothing can be said about it after the {DB,lock} is unlocked.)

From a separate discussion:
> This is mostly about preventing concurrent operations on the same container/pod - ensuring we don't call stop until the container has finished starting, etc.

This feels a bit as if there should be explicit “not running/starting/started” states in the DB, rather than “not running/not running but exclusively locked/started“.  OTOH, we would still need a mechanism for one or more processes to wait until the “starting” state transitions to “started”, and poll loops would certainly work, but proper blocking might actually be more difficult to do than mutexes – ISTM we would need condition variables integrated with the BoltDB lock, otherwise we could get missed wakeups.

---

> This should be fine as the only time we take two container locks simultaneously internally is when one container depends on another, and we have a number of checks to ensure that a container cannot depend on a removed container and a dependency container cannot be removed before the containers that depend on it - so we shouldn't ever reuse a lock in that case.

What worries me about this design is that “cannot depend on a removed container” is no longer a sufficient guarantee, or something that can be reasoned about.

Maybe the scenario below is inherently impossible, e.g. because `syncContainer` only ever happens with exactly one lock held. It’s just not visible at this level.

---

[I don’t know how that code works, so I’ll just talk about lock ordering] Consider a container “dependency“ A→B, with locks 0 and 1, and some consumer X locking both A and B, in the correct order (0→1), then unlocking them, blocking for $whatever [and expecting to re-lock them, again in the correct order (0→1), afterwards, to do another check]. At the time X blocks for $whatever, all invariants WRT removed containers and dependencies are satisfied.

Meanwhile, another podman command deletes containers A and B (in whatever is the right dependency-observant order), and deallocates the locks 0 and 1. Both before and after this deletion happens, all invariants WRT removed containers and dependencies are satisfied.

Now, some entirely different process allocates the locks 0 and 1 for another set of containers C→D, but with the reverse order (1→0) for some reason [perhaps the lock 0 was allocated for a third unrelated purpose for a short while, we always allocate the lowest-numbered one after all], and tries to lock them in the correct order for C→D, i.e. 1→0. Again, all invariants WRT containers C,D are satisfied. It manages to grab lock 1.

Meanwhile, consumer X wakes up, and wants to lock A→B in the correct order for A→B, i.e. 0→1, and manages to grab lock 0.  The two processes now deadlock.